### PR TITLE
Issue #13626: Reuse in metadata module scrapper all methods to get data from javadoc and reflection only

### LINKED ISSUE #13626: Rewrite metadata module
One of the goals of this Xdoc automation project is to gut the Javadocs. The problem is that they can not be deleted immediately. The metadata scraper relies on these Javadocs to generate the XML metadata files:
https://github.com/checkstyle/checkstyle/blob/98e74d072a8c20ab4cf9f0fdcb35c3cbdb4e7aaf/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java#L48-L52

After the properties macro (https://github.com/checkstyle/checkstyle/pull/13407) is merged, we will have all the functionality to make this rewrite possible - properties macro,  parent module macro, violation messages macro. Additionally, a module will have to be written to get the description from the Xdoc.

All this functionality can be moved to a shared module where the macros and the metadata generator can be used.

After completion of this issue, the Javadocs of all modules can be removed.

### PR DESCRIPTION
**Pull request by stoyanK7**
Resolves #13626

### CODE REVIEW

**Comment by nrmancuso** [comment]
@stoyanK7 let's rebase and prepare for review :)

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 1]
```diff
@@ -22,25 +22,57 @@
...
+    /** Regular expression for detecting ANTLR tokens(for e.g. CLASS_DEF). */
+    private static final Pattern TOKEN_TEXT_PATTERN = Pattern.compile("([A-Z_]{2,})+");
+
+    /**
+     * Modules whose metadata is generated by functionality shared with SiteUtil, not
+     * by JavadocMetadataScraper.
+     * Until <a href="https://github.com/checkstyle/checkstyle/issues/13626">#13626</a>
+     */
+    private static final Set<String> MODULES_GENERATED_BY_NEW_FUNCTIONALITY = Set.of(
+        "TodoCommentCheck.java"
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:73]: to avoid dealing with such list, that cause conflicts .... annoying.
can we have some code that check if template has "<macros " keyword we activate few functionality ?

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 2]
```diff
@@ -74,8 +107,190 @@ public static void generate(String path, OutputStream out, String... moduleFolde
...
+        final List<File> moduleFilesForNewMetadata = moduleFiles.stream().filter(file -> {
+            final String fileName = file.getName();
+            return MODULES_GENERATED_BY_NEW_FUNCTIONALITY.contains(fileName);
+        }).collect(Collectors.toList());
+        moduleFiles.removeAll(moduleFilesForNewMetadata);
 
         checker.process(moduleFiles);
+
+        //////////////////////////////
+        // Plan:
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:119]: please move this block of code to new method.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:73]: I do not care about performance here, so simple load of while file as sting to memory  and doing "contains" it is good.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 3]
```diff
@@ -74,8 +107,190 @@ public static void generate(String path, OutputStream out, String... moduleFolde
...
+                moduleDetails.addToProperties(propertyDetails);
+            }
+
+            try {
+                XmlMetaWriter.write(moduleDetails);
+            }
+            catch (TransformerException | ParserConfigurationException ex) {
+                throw new IllegalStateException(
+                                "Failed to write metadata into XML file for module: "
+                                        + "todoComment", ex);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:199]: Put here name of module

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 4]
```diff
@@ -74,8 +107,190 @@ public static void generate(String path, OutputStream out, String... moduleFolde
...
+            moduleDetails.addToViolationMessages(new ArrayList<>(messageKeys));
+
+            final String className = getModuleName(file);
+            final Set<String> properties = SiteUtil.getPropertiesForDocumentation(clss, instance);
+            final Map<String, DetailNode> javadocs = SiteUtil
+                    .getPropertiesJavadocs(properties, className, file);
+            final DetailNode moduleJavadoc = javadocs.get(className);
+            moduleDetails.setDescription(constructSubTreeText(moduleJavadoc));
+
+            final ModulePropertyDetails modulePropertyDetails = new ModulePropertyDetails();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:168]: I don't see usage of it.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 5]
```diff
@@ -74,8 +107,190 @@ public static void generate(String path, OutputStream out, String... moduleFolde
...
+    private static ModuleType getModuleType(String moduleName) {
+        final ModuleType result;
+        if (moduleName.endsWith("FileFilter")) {
+            result = ModuleType.FILEFILTER;
+        }
+        else if (moduleName.endsWith("Filter")) {
+            result = ModuleType.FILTER;
+        }
+        else {
+            result = ModuleType.CHECK;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:291]: Are you sure in this ? 
All Checks has name suffix "Check". 
There are more modules like Holder, that we s bad hack but still we have them.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 6]
```diff
@@ -74,8 +107,190 @@ public static void generate(String path, OutputStream out, String... moduleFolde
...
+                moduleName = moduleName
+                        .substring(0, moduleName.length() - checkModuleExtension.length());
+            }
+            moduleDetails.setName(moduleName);
+
+            // TODO: Move to shared module, because that MacroExecutionException doesn't belong here
+            final Object instance = SiteUtil.getModuleInstance(moduleName);
+            final Class<?> clss = instance.getClass();
+            final String fullyQualifiedName = clss.getName();
+            moduleDetails.setFullQualifiedName(fullyQualifiedName);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:149]: Instead of setters, can we use all-argumets constructor of module details?
And move all calculations of value to special methods.

So our code with look:
```
Final String name = getFQN(....)
Object parent = get parent(...)
.....
ModuleDetails moduleDetails = new ModuleDetails(name, parent,.....)
```

Same for ModulePropertyDetails.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 5 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:291]: Holders are still checks.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:291]: Ok, we will deal with this later on

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/AnnotationLocationCheck.xml [hunk 7]
```diff
@@ -38,24 +38,24 @@
...
                       type="boolean">
-               <description>Allow single parameterless
- annotation to be located on the same line as target element.</description>
+               <description>Allow single parameterless annotation to be located on the same line as
+ target element.</description>
             </property>
             <property default-value="CLASS_DEF,INTERFACE_DEF,PACKAGE_DEF,ENUM_CONSTANT_DEF,ENUM_DEF,METHOD_DEF,CTOR_DEF,VARIABLE_DEF,RECORD_DEF,COMPACT_CTOR_DEF"
                       name="tokens"
-                      type="java.lang.String[]"
+                      type="[Ljava.lang.String;"
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/AnnotationLocationCheck.xml:58]: Not good, please fix

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 6 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 4 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Removed.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 3 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 2 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 1 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Uhh, I got something working but metadata generation is 10 seconds slower. Is that alright or do we aim lower?

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/AnnotationLocationCheck.xml (hunk 7 above)
Comment (stoyanK7) [reply src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/AnnotationLocationCheck.xml:58]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 8]
```diff
@@ -105,4 +133,192 @@ private static List<File> getTargetFiles(String path, String... moduleFolders)
...
+     */
+    private static String getValidationType(String property) {
+        final String validationType;
+        if (SiteUtil.TOKENS.equals(property) || SiteUtil.JAVADOC_TOKENS.equals(property)) {
+            validationType = "tokenSet";
+        }
+        else {
+            validationType = null;
+        }
+        return validationType;
```
Comment (stoyanK7) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:238]: We will have to transfer the `validationType` to the Javadoc. What about a Javadoc param?
>  * Validation type is {code tokenTypesSet}.
```
    /**
     * description...
     *
     * @param allow User's value of allowSamelineSingleParameterlessAnnotation.
     * @validationType tokenTypesSet
     * @since 6.1
     */
    public final void setAllowSamelineSingleParameterlessAnnotation(boolean allow) {
        allowSamelineSingleParameterlessAnnotation = allow;
    }
```

Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:238]: Doesn't have to be in the scope of this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 9]
```diff
@@ -22,21 +22,36 @@
...
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.TransformerException;
+
+import org.apache.maven.doxia.macro.MacroExecutionException;
```
Comment (stoyanK7) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:41]: It isn't ideal that the metadata module uses the site module. I think it's better if the opposite is the case - the site module uses the metadata module. Shall we deal with this in a follow-up issue?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:41]: We will refactor later on to move methods around to make it better design wise.
For now let's focus on functionality to unblock ability to remove properties from javadoc

**Comment by romani** [comment]
I think PR is ready to not have commit for disablement of CIs, please remove.

Please rephrase commit message, as you do not rewrite but you change a way we get data and we will do this very gradually.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: @stoyanK7 , what about this ?

Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java]: Done as well.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:262]: Please create separate issue on this and make example in read properly and validation type to let me understand

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:262]: For now I think this hardcod is enough.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 10]
```diff
@@ -105,4 +157,192 @@ private static List<File> getTargetFiles(String path, String... moduleFolders)
...
+    }
+
+    /**
+     * Get property description from property javadoc.
+     *
+     * @param property property name.
+     * @param propertyJavadoc property javadoc.
+     * @return property description.
+     */
+    private static String getPropertyDescription(String property, DetailNode propertyJavadoc) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:272]: I am not sure why this method is specific for metadata.
I think there should single method that does it for xdoc and meta and meta generation just reuse it 

If agree please create separate issue to improve in separate PR

**Review by romani — APPROVED** [review]
Ok to merge to unblock massive removal of javadoc content.
We will improve in separate PRs

**Comment by romani** [comment]
@stoyanK7 , please make CI green.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java [hunk 11]
```diff
@@ -1110,12 +1117,119 @@ public static String getLinkToDocument(String moduleName, String document)
...
+            final List<DetailNode> descriptionNodes = getDescriptionNodes(javadoc);
+            Lists.reverse(descriptionNodes).forEach(stack::push);
+            final StringBuilder result = new StringBuilder(1024);
+            while (!stack.isEmpty()) {
+                detailNode = stack.pop();
+
+                Lists.reverse(Arrays.asList(detailNode.getChildren())).forEach(stack::push);
+
+                final String childText = detailNode.getText();
+                // Regular expression for detecting ANTLR tokens(for e.g. CLASS_DEF).
```
[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:1219]: Can we just use https://github.com/checkstyle/checkstyle/blob/54616f662130aa0ca5afe45f72d36cdefb78a781/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java#L96 to get a list of actual tokens and check that?

File: pom.xml [hunk 12]
```diff
@@ -860,6 +860,7 @@
                     <exclude>
                       com.puppycrawl.tools.checkstyle.api.AutomaticBean*
                     </exclude>
+                    <exclude>com.puppycrawl.tools.checkstyle.meta.MetadataGeneratorUtil</exclude>
```
[the remark this answers is not in the document]
Comment (romani) [reply pom.xml:863]: This is jacoco coverage.
The only time it makes sense if we ever move all functionality to separate project and we remote xdoc files from git completely, as we will never see how content looks like before html generations.
Before this time, all such methods should not have to work on all possible cases, code should serve our current cases of   Checks and xdoc/meta. So coverage has no much meaning.
Do we need such issue ? As we do not even has plan to move it out

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 9 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:41]: Let's collect ideas at https://github.com/checkstyle/checkstyle/issues/13699 to improve in separate PRs

File: pom.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (romani) [reply pom.xml:863]: If you agree, please copy my comment or part of it to description at https://github.com/checkstyle/checkstyle/issues/13426 as requirements

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 13]
```diff
@@ -105,4 +157,192 @@ private static List<File> getTargetFiles(String path, String... moduleFolders)
...
+            final String validationType = getValidationType(property);
+
+            result.add(new ModulePropertyDetails(property, type, defaultValue,
+                    validationType, description));
+        }
+        return result;
+    }
+
+    /**
+     * Get validation type for the given property.
```
[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:249]: Actual type is int[] but business logic wise it is int values from "enum"-like class of tokens. So this helps plugins to know how to validate content. Ideally it should be hinted by source code but we never end up doing it explicitly.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:262]: @stoyanK7 , your example is very confusing.
We use validation type for more complicated types, not a booleans 
https://github.com/checkstyle/checkstyle/blob/f58e45eb77487425249b7b0cc2b8bebe54662b51/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/AnnotationLocationCheck.xml#L59

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:262]: Only tokens https://github.com/search?q=repo%3Acheckstyle%2Fcheckstyle+validation-type&type=code&p=1

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 10 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:272]: If you agree add it to https://github.com/checkstyle/checkstyle/issues/13699

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/ModuleDetails.java [hunk 14]
```diff
@@ -47,6 +47,34 @@ public final class ModuleDetails {
     /** Type of the module(check/filter/filefilter). */
     private ModuleType moduleType;
 
+    /** No-argument constructor. */
+    public ModuleDetails() {
+        // empty constructor
+    }
```
[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/ModuleDetails.java:53]: Compile error https://stackoverflow.com/a/5286532/1015848
We still have old code of metadata that use default ctor.

We are not rewriting metadata, we migrate it to use new method to get data that are based on reflection and javadoc

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 10 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:272]: I think there should be separate methods for xdoc and metadata. Or at least single method with different output depending on who/what is calling it.

Metadata gets text as-is and doesn't care about any processing.

Xdoc requires to 
- convert `{@code something}` to `<code>something</code>`
- Convert links to relative ones-  `https://checkstyle.org/checks/block/xxx.html` -> `../block/xx.html`, for example

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/ModuleDetails.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/ModuleDetails.java:53]: There is usage of it in `JavadocMetadataScraper`

https://github.com/checkstyle/checkstyle/blob/dbcbd67be9cd897d7fb46c7b7bb9cf448276f8fc/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java#L192

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/ModulePropertyDetails.java [hunk 15]
```diff
@@ -40,6 +40,29 @@ public final class ModulePropertyDetails {
     /** Description of property. */
     private String description;
 
+    /** No-argument constructor. */
+    public ModulePropertyDetails() {
+        // empty constructor
+    }
```
[the remark this answers is not in the document]
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/ModulePropertyDetails.java:46]: Same as above

https://github.com/checkstyle/checkstyle/blob/dbcbd67be9cd897d7fb46c7b7bb9cf448276f8fc/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java#L295

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java [hunk 16]
```diff
@@ -105,4 +157,192 @@ private static List<File> getTargetFiles(String path, String... moduleFolders)
...
+                moduleType, propertiesDetails, new ArrayList<>(messageKeys));
+    }
+
+    /**
+     * Get module type(check/filter/filefilter) based on module name.
+     *
+     * @param moduleName module name.
+     * @return module type.
+     */
+    private static ModuleType getModuleType(String moduleName) {
```
[the remark this answers is not in the document]
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:203]: You mean make the one in `JavadocMetadataScraper` `public`?

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 10 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:272]: But still data creation should be single method and xdoc just do only such required formatting. If you agree just put this to separate issue that we have for design improvement as item.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:203]: I have no strong feelings either way, but we should consider some way to avoid massive copy paste

**Comment by rnveach** [comment]
@stoyanK7 Since we are re-writing meta, did you verify if any of your changes will break eclipse-cs or sonar?

**Comment by romani** [comment]
We are not rewriting, we are using different approach to get data for metadata files that we generate. No changes in files, it means it works as before.

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java (hunk 8 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java:262]: Moving to https://github.com/checkstyle/checkstyle/issues/13722

**Comment by romani** [comment]
very sad that me failed to merge this, issue is still open, I hope we can come back to it.
